frontend/src/hooks/useFormData/index.tsx

```diff
@@ -43,43 +52,48 @@ const useFormData = (): [FormData, Meta, Helpers] => {
         };
     };
 
-    const validateForm = () => {
+    const validateForm = (): ValidFormData | ErrorMessage.OnSubmit => {
         const stateUpdates = objectKeys(formData).reduce<(() => void)[]>(
             (updates, field) => {
                 const fieldError = formData[field].validate();
-                if (fieldError) {
+                const isMissing = isFieldMissing(field, formData);
+                const error: FieldError =
+                    fieldError || (isMissing && ErrorMessage.Required) || null;
+
+                if (fieldError || isMissing) {
                     updates.push(() =>
                         setFormData((prevData) => ({
                             ...prevData,
-                            [field]: { ...prevData[field], error: fieldError },
+                            [field]: { ...prevData[field], error },
                         }))
                     );
                 }
+
                 return updates;
             },
             []
         );
 
-        if (stateUpdates.length) {
+        if (noErrorsDetected(stateUpdates, formData)) {
+            return formData;
+        } else {
             stateUpdates.forEach((update) => update());
             return ErrorMessage.OnSubmit;
-        } else {
-            return null;
```

Comment:
First issue with this PR is here.

Background, I inverted `validateForm` to return the equivalent of an Either (just now realizing that when you were talking about doing a Maybe/Either in an earlier PR, you just meant `X | Y`, not actually implementing a type called `Maybe`/`Either`? 🤦‍♀️ ). Anyway, returns either: valid form data, or an error message.

I did that to help with a second change, which is returning `formData` after passing it through  `noErrorsDetected`, which is a helper function that coerces `formData` from type `FormData` into`ValidFormData`. `ValidFormData` is the same thing, except it does not allow `null` for required fields.

This lets me transform further into the `GameReportPayload`, which requires certain previously nullable values to now be non-nullable values.

Basically my problem with it: it opens us up to a `400` error from the server if the validation were to not properly confirm non-null values, since I'm just blanket telling Typescript to believe me.

Reply:
> a pattern called a Type Guard, where you narrow a type based on runtime checking of the value semantics. There isn't really anything wrong with this approach. It does have the downside you've called out, which is that your type-checking is vulnerable now to runtime logic, but that doesn't make it an invalid technique

TIL, I didn't know there was a name for this, that's really cool and I feel better about what I did!

> If you do keep the type guard, I would perhaps consider ditching the helper function for clarity. It's probably simpler to parse this as `if (stateUpdates.length) return formData as ValidFormData`

This is much simpler and easier to see clearly what's going on when it's not moved off into a function, updated ✔️ 

> If you really hate the type guard, then you need to construct the `ValidFormData` explicitly while performing the appropriate `null` checking

Now that you explained what the pattern is I don't hate it anymore! I feel like I'm making the opposite decision here to the mapping from `ValidFormData` to `GameReportPayload`, but in that case the tradeoff is only more lines of code, while in this case it would mean more lines of code AND doing null checking twice, and it just doesn't feel worth it.